### PR TITLE
Reactive node replacement: AWS-IPI

### PR DESCRIPTION
**Pull request by PrasadDesala**
Signed-off-by: Prasad Desala <tdesala@redhat.com>

### CODE REVIEW

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        # Check the pods should be in running state
+        all_pod_obj = pod.get_all_pods(wait=True)
+        for pod_obj in all_pod_obj:
+            wait_for_resource_state(
+                resource=pod_obj, state=constants.STATUS_RUNNING, timeout=300
+            )
+
+        # Check basic cluster functionality by creating resources
+        # (pools, storageclasses, PVCs, pods - both CephFS and RBD),
+        # run IO and delete the resources
```
Comment (hnallurv) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: how are you making sure that the dc pods running on the failed node are restarted on new node and can continue running the IOs?

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        )
+        log.info(
+            f"{osd_node_obj.name} associated machineset is {machineset_name}"
+        )
+
+        # Induce failure
+        nodes.stop_nodes([osd_node_obj], wait=True)
+        log.info(f"Successfully powered off node: {osd_node_obj.name}")
+
+        # TODO - Network failure as one more failure type
```
Comment (hnallurv) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: will there be a separate pr for this?

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+
+@ignore_leftovers
+@tier4
+@tier4b
+@aws_platform_required
+@ipi_deployment_required
+class TestNodeReplacement(ManageTest):
+    """
+    Knip-894 Node replacement - AWS-IPI-Reactive
```
Comment (hnallurv) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: please add polarion test id

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            f"Successfully labeled {new_spun_node} with OCS storage label"
+        )
+
+        # Check the pods should be in running state
+        all_pod_obj = pod.get_all_pods(wait=True)
+        for pod_obj in all_pod_obj:
+            wait_for_resource_state(
+                resource=pod_obj, state=constants.STATUS_RUNNING, timeout=300
+            )
+
```
Comment (hnallurv) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Need a check on whether the IOs from pods running on surviving worker nodes are not disturbed because of replacement.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        # Check basic cluster functionality by creating resources
+        # (pools, storageclasses, PVCs, pods - both CephFS and RBD),
+        # run IO and delete the resources
+        self.sanity_helpers.create_resources(pvc_factory, pod_factory)
+
+        # Perform cluster and Ceph health checks
+        self.sanity_helpers.health_check()
```
Comment (hnallurv) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: include post replacement data integrity check

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 3 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: This test script can handle 1AZ, 3AZ and failures of combination power off and network failure.
Unsure on how to add the test case ID here since in polarion we only have a single generic test case.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 2 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: I will try to add it in the same PR. Didn't had a setup to try it out so mentioned TBD for now.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 1 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: This is not happening and seems to be a bug. I am going to file one, so didnt implement it yet.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 4 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Yes, I have planned for it. The dc pods are not failing over to the other node and stuck with that. 
I'll add this verification once that issue is fixed.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+        """
+        # Get worker nodes
+        initial_nodes = get_worker_nodes()
+        log.info(f"Current available worker nodes are {initial_nodes}")
+
+        # Get the osd associated node
+        osd_pods_obj = pod.get_osd_pods()
+        osd_node_obj = pod.get_pod_node(random.choice(osd_pods_obj))
+        log.info(f"Selected OSD is {osd_node_obj.name}")
```
Comment (fbalak) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]:
```suggestion
        log.info(f"Selected OSD node is {osd_node_obj.name}")
```

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 3 above)
Comment (ebenahar) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: This is indeed a new thing. We are not used to creating test permutations based on IPI or UPI.
@PrasadDesala please decide whether to create a polarion TC for each deployment type based on if the test steps are fundamentally different between IPI and UPI

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    @pytest.fixture(autouse=True)
+    def init_sanity(self):
+        """
+        Initialize Sanity instance
+
+        """
+        self.sanity_helpers = Sanity()
+
+    def test_node_replacement_reactive_aws_ipi(
+            self, nodes, pvc_factory, pod_factory,
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: delete extra indentation

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            dc_pod_factory
+    ):
+        """
+        Knip-894 Node replacement - AWS-IPI-Reactive
+
+        """
+        # Get worker nodes
+        initial_nodes = get_worker_nodes()
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: please try using nodes.get_typed_nodes()
I think we should use only this function as it is generic for both workers and masters

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        log.info(f"Current available worker nodes are {initial_nodes}")
+
+        # Get the osd associated node
+        osd_pods_obj = pod.get_osd_pods()
+        osd_node_obj = pod.get_pod_node(random.choice(osd_pods_obj))
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Wondering if the test will have different results with randomly picking OSD, between when the picked one resides in an AZ that has another node with OSD and when the picked one resides on an AZ alone

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            else:
+                pod.run_io_in_bg(dc_rbd, expect_to_fail=False, fedora_dc=True)
+            dc_rbd_pod_obj.append(dc_rbd)
+
+        dc_cephfs_pod_obj = []
+        for node in initial_nodes:
+            # Create app pods on all the nodes
+            dc_cephfs = dc_pod_factory(
+                interface=constants.CEPHFILESYSTEM, node_name=node)
+            if node == osd_node_obj.name:
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: please set expect_to_fail prior to calling run_io_in_bg() and then pass this var to the function so it will be called only once

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        )
+
+        # Induce failure
+        nodes.stop_nodes([osd_node_obj], wait=True)
+        log.info(f"Successfully powered off node: {osd_node_obj.name}")
+
+        # TODO - Network failure as one more failure type
+
+        # Add annotation to the failed node
+        annotation = "machine.openshift.io/exclude-node-draining=''"
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: will be good to have this functionality in a function in machine.py

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 1 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Now we are checking whether DC pods are in running state or not after failure.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 2 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Done. Added network failure.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 3 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Created polarion test cases for each failure and volume combination .

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 4 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 5 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 7 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 8 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 9 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Now this logic for selection of OSDs has been changed.

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 10 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py (hunk 11 above)
Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Done

**Comment by PrasadDesala** [comment]
Closing this old PR. Will open a new one.
